### DISCUSSION TITLE
db 동시성 dirty read

### DISCUSSION THREAD

**Discussion by LEEHYUNDONG** [CS-studi/CS-study#22] [Q&A] [answered]
@guswns3371 

```
1. 트랜잭션1이 데이터a를 갱신함
2. 트랜잭션2가 "갱신된 데이터a"를 읽음
3. 트랜잭션1이 rollback되어 갱신이 무효화됨

트랜잭션2는 무효화된 "갱신된 데이터a"를 읽게 된다.
```

```
원래 데이터 a가 존재
1. 트랜잭션1이 데이터a를 b로 갱신함 
2. 트랜잭션2가 데이터를 읽음, "데이터 b"를 읽음
3. 트랜잭션1이 rollback되어 갱신이 무효화됨
4. 트랜잭션2가 데이터를 다시 읽음, "데이터 a"가 읽힘
```

현준씌 이거 적어주신 예시에서 말이 조금 헷갈리는데요.
결국 트랜잭션 2데이터 갱신 부분이 rollback되서 맨 아랫줄에 결론에 ___트랜잭션2는 "데이터a"를 읽게 된다___ 이 뜻이고 아래처럼 해석하면 되는거죠?

**Comment by guswns3371** [comment, marked as answer]
트랜잭션 2는 데이터를 다시 읽지 않습니다
트랜잭션 1이 데이터를 a->b 로 갱신하면 트랜잭션2가 읽은 데이터는 b입니다.

이제 트랜잭션 1이 정상적으로 종료하면 트랜잭션2는 올바른 데이터를 읽은거지만,

트랜잭션 1이 비정상종료가 되면 데이터는 다시 b에서 a로 롤백됩니다. 결과적으로 트랜잭션2가 읽은 데이터는 b인거죠.(잘못된 데이터를 읽은 것)

**Reply by LEEHYUNDONG** [reply]
트랜잭션2가 읽었을때 dirty read문제가 발생했고 그래서 롤백되는거로 알고 있는데 잘못된거면 알려주세요 !ㅎㅎ

http://blog.skby.net/dirty-read/

**Reply by guswns3371** [reply]
[dirty read1](http://blog.skby.net/dirty-read/#:~:text=%EB%B3%80%EA%B2%BD%20%ED%9B%84%20%EC%95%84%EC%A7%81%20Commit%20%EB%90%98%EC%A7%80%20%EC%95%8A%EC%9D%80%20%EA%B0%92%20%EC%9D%BD%EA%B3%A0%2C%20Rollback%20%ED%9B%84%EC%9D%98%20%EA%B0%92%EC%9D%84%20%EB%8B%A4%EC%8B%9C%20%EC%9D%BD%EC%96%B4%20%EC%B5%9C%EC%A2%85%20%EA%B2%B0%EA%B3%BC%20%EA%B0%92%EC%9D%B4%20%EC%83%81%EC%9D%B4%ED%95%9C%20%ED%98%84%EC%83%81) :  변경 후 아직 Commit 되지 않은 값 읽고, Rollback 후의 값을 `다시 읽어` 최종 결과 값이 상이한 현상 (http://blog.skby.net/dirty-read/)

---

[dirty read2](https://www.cubrid.com/tutorial/3794141#:~:text=Dirty%20Read%EB%9E%80%20T1%20%ED%8A%B8%EB%9E%9C%EC%9E%AD%EC%85%98%EC%97%90%EC%84%9C%20%EC%9D%BD%EA%B8%B0%ED%95%98%EA%B3%A0%20%EC%8B%A4%ED%96%89%EC%9D%B4%20%EB%81%9D%EB%82%98%EC%A7%80%20%EC%95%8A%EC%9D%80%20%EC%83%81%ED%83%9C%EC%97%90%EC%84%9C%20T2%20%ED%8A%B8%EB%9E%9C%EC%9E%AD%EC%85%98%EC%97%90%20%EC%9D%98%ED%95%B4%20%EB%B3%80%EA%B2%BD%EB%90%9C%20%EC%82%AC%ED%95%AD%EC%9D%84%20%EB%B3%B4%EA%B2%8C%20%EB%90%98%EB%8A%94%20%EA%B2%BD%EC%9A%B0%EB%A5%BC%20dirty%20read%EB%9D%BC%EA%B3%A0%20%ED%95%9C%EB%8B%A4%2C%20%EB%A7%8C%EC%95%BD%20T2%20%ED%8A%B8%EB%9E%9C%EC%9E%AD%EC%85%98%EC%97%90%EC%84%9C%20%EA%B7%B8%20%EB%B3%80%EA%B2%BD%20%EC%82%AC%ED%95%AD%EC%9D%84%20rollback%ED%95%98%EB%A9%B4%20%EA%B7%B8%20%EB%8D%B0%EC%9D%B4%ED%84%B0%EB%A5%BC%20%EC%9D%BD%EC%9D%80%20T1%20%ED%8A%B8%EB%9E%9C%EC%9E%AD%EC%85%98%EC%9D%80%20Dirty%20%EB%8D%B0%EC%9D%B4%ED%84%B0%EB%A5%BC%20%EA%B0%80%EC%A7%80%EA%B3%A0%20%EC%9E%88%EB%8A%94%20%EA%B2%83%EC%9D%B4%EB%8B%A4) : Dirty Read란 T1 트랜잭션에서 읽기하고 실행이 끝나지 않은 상태에서 T2 트랜잭션에 의해 변경된 사항을 보게 되는 경우를 dirty read라고 한다, 만약 T2 트랜잭션에서 그 변경 사항을 rollback하면 그 데이터를 읽은 T1 트랜잭션은 Dirty 데이터를 가지고 있는 것이다.

[dirty read3](https://www.tutorialspoint.com/what-is-dirty-read-in-a-transaction-dbms#:~:text=Dirty%20read%20is%20a%20read%20of%20uncommitted%20data.%20If%20a%20particular%20row%20is%20modified%20by%20another%20running%20application%20and%20not%20yet%20committed%2C%20we%20also%20run%20an%20application%20to%20read%20the%20same%20row%20with%20the%20same%20uncommitted%20data.%20This%20is%20the%20state%20we%20say%20it%20as%20a%20dirty%20read) : `Dirty read is a read of uncommitted data`. If a particular row is modified by another running application and not yet committed, we also run an application to read the same row with the same uncommitted data. This is the state we say it as a dirty read. 

[quora: dirty read4](https://www.quora.com/What-is-a-dirty-read-in-database-systems)
![image](https://user-images.githubusercontent.com/44316546/161093156-8700ae1b-b070-4d21-b5c8-83b7cba83777.png)

```java
1. Alice and Bob start two database transactions.
2. Alice modifies the title of a given post record.
3. Bob reads the uncommitted post record.
4. If Alice commits her transaction, everything is fine. 
But if Alice rolls back, then Bob will see a record version which no longer exists in the database transaction log.
```

제가 작성한 예시가 quora의 예시와 동일한거같아요. (근데 또 외국인이 말했다 해서 정답은 아니니까..)

## [mysql : dirty read](https://dev.mysql.com/doc/refman/8.0/en/glossary.html#:~:text=dirty%20read,in%20the%20meantime.)

> An operation that retrieves unreliable data, data that was updated by another transaction but not yet committed. It is only possible with the isolation level known as read uncommitted.
>  This kind of operation does not adhere to the ACID principle of database design. It is considered very risky, because the data could be rolled back, or updated further before being committed; then, the transaction doing the dirty read would be using data that was never confirmed as accurate.
> Its opposite is consistent read, where InnoDB ensures that a transaction does not read information updated by another transaction, even if the other transaction commits in the meantime.

mysql 공식 페이지에서 설명하는 dirty read란 `the transaction doing the dirty read would be using data that was never confirmed as accurate` (`아직 확실하게 confirmed되지 않은 데이터 = commit 되기 전의 데이터`를 읽었기 때문에 발생하는 문제) 라고 하네요.

그리고 `Its opposite is consistent read` (dirty read의 반대 개념 = 일관성 있는 read) 라고 말하는걸 보아하니 데이터의 일관성이 깨지는 문제 중 하나가 dirty read 문제인거 같아요. 

## [Microsoft](https://docs.microsoft.com/en-us/sql/odbc/reference/develop-app/transaction-isolation-levels?view=sql-server-ver15#:~:text=Dirty%20Reads%20A,to%20have%20existed.)

> Dirty Reads A dirty read occurs when a transaction reads data that has not yet been committed. For example, suppose transaction 1 updates a row. Transaction 2 reads the updated row before transaction 1 commits the update. If transaction 1 rolls back the change, transaction 2 will have read data that is considered never to have existed.

`transaction 2 will have read data that is considered never to have existed.` 에서 `will have` 라고 말하는걸 보아하니 트랜잭션2가 데이터를 `다시` 읽는게 아니라 `읽었던` 데이터가 유효하지 않은 데이터인거 같아요.

---

## [stackoverflow : dirty read와 non-repeatable read의 차이점](https://stackoverflow.com/questions/18297626/difference-between-non-repeatable-read-vs-dirty-read#:~:text=Dirty%20Reads%20occur,be%20non%2Drepeatable.)

> Dirty Reads occur when one transaction reads data written by another, uncommitted, transaction. The danger with dirty reads is that the other transaction might never commit, leaving the original transaction with "dirty" data.

`leaving the original transaction with "dirty" data.` 이부분 보시면 data를 읽은 트랜잭션이 다시 data를 읽는 건 아닌듯 해요.

```
원래 데이터 a가 존재
1. 트랜잭션1이 데이터a를 b로 갱신함 (아직 commit되기 전)
2. 트랜잭션2가 데이터를 읽음, "데이터 b"를 읽음 (commit되기 전의 데이터를 읽음)
3. 트랜잭션1이 rollback되어 갱신이 무효화됨
4. 트랜잭션2이 가지고 있는 데이터는 dirty한 데이터 (무효화된 데이터)
```

**Reply by LEEHYUNDONG** [reply]
우선 작성해주신거 잘읽었습니다.🙂

제 질문은 제안해주셨던 예시가 틀렸다는게 아니라,

> `트랜잭션2는 무효화된 "갱신된 데이터a"를 읽게 된다.`
1. 마지막 이 결론이 무효화된 갱신된 데이터 = 원래 데이터이냐는 질문이었습니다.

---

답변으로 주신 

> 트랜잭션 2는 데이터를 다시 읽지 않습니다
트랜잭션 1이 데이터를 a->b 로 갱신하면 트랜잭션2가 읽은 데이터는 b입니다.

2. 위 내용은 읽지 않는다이고, 원래 있던 예시에는 무효화된 "갱신된 데이터a"를 읽게된다라고 써놓으셔서, 이 예시에서 트랜잭션 2가 읽기 연산을 한번 더 수행하는 게 아닌가 했습니다. 그래서 [참고](http://blog.skby.net/dirty-read/) 를 달아놓고 이 예시와 동일한 예시인 건지 여쭤본거였습니다!


결론은 dirty read는 `커밋되지 않은 읽기 상황`인 것이고, rollback 여부 + dirty read가 발생한 이후에는 데이터를 읽는지 읽지 않는지는 중요하지 않다로 이해하면 될 거 같네요. (rollback이 되지 않으면 non-repeatable read, rollback이 되지 않으면 repeatable read)

감사합니다!

**Reply by LEEHYUNDONG** [reply]
결론적으로 현준님이 말씀하시는거 이해했습니다!
저는 위 예시가 옳다고 생각합니다.
오늘도 화이팅해봐요🙏

**Reply by guswns3371** [reply]
👍👍👍
`트랜잭션2는 무효화된 "갱신된 데이터a"를 읽게 된다.`가 아닌
`
트랜잭션2는 무효화된 "갱신된 데이터a"를 가지게 된다.`로 변경하면 의미 전달이 확실해질거 같네요!
